File: src/systemd.c
#include <mruby.h>
#include <mruby/string.h>
#include <systemd/sd-daemon.h>

static mrb_value
mrb_c_systemd_notify(mrb_state *mrb, mrb_value self)
{
	mrb_int unset_environment;
	mrb_value state;
	mrb_get_args(mrb,"iS", &unset_environment, &state);
	const char *cstate;
	int cunset_environment;
	cstate = mrb_string_value_cstr(mrb, &state);
	cunset_environment = (int) unset_environment;
	return mrb_fixnum_value(sd_notify(cunset_environment,cstate));
}

void
mrb_systemd_gem_init(mrb_state* mrb) {
  struct RClass *csystemd;
  csystemd = mrb_define_class(mrb, "CSystemd",  mrb->object_class);
  mrb_define_class_method(mrb, csystemd, "notify", mrb_c_systemd_notify, MRB_ARGS_REQ(2));
}

void
mrb_systemd_gem_final(mrb_state* mrb) {
  /* finalizer */
}

File: mrblib/systemd.rb
class Systemd
	class << self
		def notify *args
			args[0].each do |k,v|
				state = "READY=#{v}" if k == :ready
				unset_environment = v if k == :unset_environment
			end if args[0]

			unset_environment ||= 1
			state             ||= "READY=1"

			unset_environment = 1 if unset_environment == true
			unset_environment = 0 if unset_environment == false

			CSystemd.notify(unset_environment,state)
		end
	end
end